Before you turn this homework in, make sure everything runs as expected. First, **restart the kernel** and then **run all cells** (in the menubar, select Kernel$\rightarrow$Restart \& Run all). If you get error messages, the grader will get the same errors and won't be able to grade your work. 

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE". You may delete the line(s) that say "**raise** NotImplementedError()" and replace them with your actual work. Write all your work in the cells provided in this notebook. **DO NOT** add or delete any cells. 

Please fill in your name in the cell below.

In [1]:
NAME = "Kevin Pham"

---

# Homework 4

In all problems below, please comment your code sufficiently well so that the grader can follow what you are doing with ease. For non-coding answers, please make sure to formulate your explanation and answers in the form of complete English sentences. It is not sufficient to just leave comments in the code (without a full sentence explanations in problems that ask for those) or vice versa (to leave full sentence explanations but not code comments). You will need both to get full credit. 

## Problem 1

(4 points) Write a function named "border" that takes as its input two integers n and m ($1 \le n,m \le 100$) and outputs a 2-dimensional NumPy array of size $(n,m)$. Your array should be filled with zeros except for the "border" (that is, the first and last column and the first and last row) which should be filled with ones. 

In [2]:
# COMPLETE
# Takes in 2 ints
# outputs a 2D array of size (nxm) 
# borders and corners should be ones

import numpy as np

def border(n, m):
    """Creates an array with shape (n, m) that is all zeros
    except for the border (i.e., the first and last rows and
    columns), which should be filled with ones."""
    # YOUR CODE HERE
    new = np.zeros((n,m)) # make a matrix of zeroes
    for i in range(n): # establish a nested loop
        for j in range(m):
            # change the border rows here
            new[0,j] = 1 # top row
            new[n-1,j] = 1 # bottom row
            
            # change the border columns here
            new[i,0] = 1 # leftmost column
            new[i,m-1] = 1 # rightmost column
               
    return new # return the edited matrix
# Test
    # raise NotImplementedError()

In [3]:
# check that your function works as expected (if it does these checks should return "True")
print(np.array_equal(border(3,3), np.array([[1, 1, 1], [1, 0, 1], [1, 1, 1]])))
print(np.array_equal(border(4,5), np.array([[1., 1., 1., 1., 1.],
       [1., 0., 0., 0., 1.],
       [1., 0., 0., 0., 1.],
       [1., 1., 1., 1., 1.]])))


True
True


## Problem 2

(a) (2 points) Suppose that x and y are both NumPy arrays of size $(1,n)$. 
Write a function that computes Pearson's sample correlation coefficient $r$ 
between the entries in x and y. 

Recall, that Pearson's sample correlation coefficient is defined as 
$$r(x,y) = \frac{\sum_{i=1}^n (x_i - \bar{x})(y_i - \bar{y})}{\sqrt{\sum\limits_{i=1}^n (x_i - \bar{x})^2 \sum\limits_{i=1}^n(y_i - \bar{y})^2}} $$

**Note:** Do *not* use the built in Numpy function ```np.corrcoef()```. Instead, please use Numpy ufuncs for sums, means, and squareroots.

In [4]:
# COMPLETE 

def pearsons_correlation(x, y):
    """Computes Pearson's correlation coefficient between vectors x and y."""
    # ESTABLISH COMPONENTS HERE
    me_x = np.mean(x)
    me_y = np.mean(y)
    
    # note: When doing component-wise multiplication (x*y) is already doing multiplication component-wise
    numer = np.sum((x-me_x)*(y-me_y)) # numerator of the pearson coeff.
    
    # components of the denominator for pearson corr.
    denom_01 = np.sum((x-me_x)**2)
    denom_02 = np.sum((y-me_y)**2)
    
    # full denominator
    denom = np.sqrt(denom_01*denom_02)
    
    pCoeff = numer/denom # compute full pearson corr.
    
    return pCoeff # return the corr.
   
    # YOUR CODE HERE
    # raise NotImplementedError()


x = np.random.random((1,10))
y = np.random.random((1,10)) 

# what is the difference between manually doing 1/len(x)*sum(x) and mean(x)?


In [5]:
# check that your function works
# if it does, the comparison below should return "True"

import numpy as np

np.random.seed(10)
n = 10
x = np.random.random((1,10))
y = np.random.random((1,10))    

round(pearsons_correlation(x,y),6) == round(np.corrcoef(x,y)[0,1],6)
# we're rounding, because the built-in NumPy method for correlation 
# may use more efficient arithmetic than we do in our home-made function. 


True

(b) (6 points) Create a NumPy array filled with uniform(0,1) random values with n rows and m columns (for the combinations of n and m listed below). In each case, find the maximum of the Pearson correlation coefficients for each of the $\binom{m}{2}$ pairs of columns in your array. 

That is, if the columns of your array are $x_1, \ldots, x_m$, find 
$$\max(r) = \max\limits_{a,b \in \{1,\ldots,m\}, a \neq b} r(x_a, x_b)$$

Enter the values you find into the table below (rounded to two after decimal digits).

In [6]:
# COMPLETE

# Find the max pearson coeff.
import numpy as np
np.random.seed(10)

def max_corr(n,m):
    array = np.random.random((n,m)) 
    """find maximum correlation between columns of (n,m) random uniform (0,1) array"""
    
    # 'm' total col. vectors and you will choose 2.
    pearson_list = [] # empty list for each pearson coeff calculated.
    
    # choosing between the COLUMNS ONLY!
    for i in range(m): # FIRST CHOICE
        for j in range(m): # SECOND CHOICE
            if(i < j): # accounts for repetitions and choosing 2 of the same col. arrays
                pearson_list.append(pearsons_correlation(array[:,i],array[:,j])) # perform each pearson coeff and queue results
                # note: array[:,i] -> Whatever row, ith column.
    return np.max(pearson_list) # return the max result from that list
                


# Testing here performing the rows/cols of the below table:
print(max_corr(10,20))
print(max_corr(10,50))
print(max_corr(10,100))
print(max_corr(100,20))
print(max_corr(100,50))
print(max_corr(100,100))

        
    # YOUR CODE HERE
    # raise NotImplementedError()

0.7339282571155442
0.8176873855394016
0.9272644039978728
0.2642408264268438
0.35895309673824183
0.3861266885311635


|  rows n | columns m  | max(r)  |
|:-:|:-:|:-:|
| 10  | 20  | |
| 10  | 50  |  |
| 10  | 100  |  |
| 100  | 20  |  |
| 100  | 50  |  |
| 100  | 100  |  |

Copy and paste this table to the Markdown cell below and complete it. Describe what you observe. Is the maximal correlation between columns (representing features here) an increasing function of n (number of rows or cases)? Is an increasing function of m (number of features)?

Since the data are entirely random, please note that any correlations you observe are entirely [spurious](https://www.tylervigen.com/spurious-correlations). 

|  rows n | columns m  | max(r)  |
|:-:|:-:|:-:|
| 10  | 20  |0.7339282571155442 |
| 10  | 50  | 0.8176873855394016 |
| 10  | 100  |0.9272644039978728  |
| 100  | 20  |0.2642408264268438  |
| 100  | 50  | 0.35895309673824183 |
| 100  | 100  |0.3861266885311635  |


It appears that whenever we have a lower 'n' with incremented 'm', we have a substantially maximal higher pearson correlation in contrast to higher 'n' with incremented 'm'. The rate at which the the maximums is also greater when we have lower 'n' in contrast to when we have higher 'n'. The maximal correlation between the columns is an increasing function of 'n'.

## Problem 3

(4 points) Suppose that $X$ is a matrix with n rows and m columns. In statistics, these matrices arise freqently when we collect (numerical) data on m variables which are each observed on n independent individuals. Suppose we denote the columns of $X = (x_1, \ldots, x_m)$. Then the covariance matrix of $X$ is defined as 

$$ \mbox{Cov}(X) = \left( \begin{array}{cccc}
Var(x_1) & Cov(x_1, x_2) & \cdots & Cov(x_1, x_m) \\
Cov(x_2, x_1) & Var(x_2) & \cdots & Cov(x_2, x_m) \\
\vdots & \vdots & \ddots & \vdots \\
Cov(x_m, x_1) & Cov(x_m, x_2) & \cdots & Var(x_m) 
\end{array} \right) $$

where Var($x_i$) is the sample variance of the entries in the $i^{th}$ column of $X$ and Cov($x_i, x_j$) ($i \neq j$) is the sample covariance of the entries in columns $i$ and $j$. 

$$ Cov(x_i, x_j) = \frac{1}{n-1}\sum\limits_{k=1}^n (x_{ki} - \bar{x}_i)(x_{kj}-\bar{x_j})$$

That is, the covariance matrix is a square $m\times m$ matrix whose diagonal entries are the sample variances of the columns of $X$ and whose off-diagonal entries are covariances between two columns of $X$, respectively. 

Write *your own* function that takes a Numpy array of shape (n,m) as its input and returns the  Covariance matrix of the array. Do not use any built-in functions to compute variance and/or covariance. Write your own functions, instead. It's ok to use NumPy routines for sums, squares, or square-roots. 

Check your work by comparing your result to that of the ```np.cov()``` function applied to the same input.

In [7]:
# COMPLETE
import numpy as np

def covariance_matrix(X): # Define the function.
    """ Finds the covariance matrix of (n,m) shaped array X"""
    
    # Traverse through each column even if it's a duplicate
    col_num = len(X[0,:]) # total number of columns (m)
    row_num = len(X[:,0]) # total number of rows per array (n)
    cov_matrix = np.zeros((col_num,col_num)) # define an empty SQUARE matrix for all computed inputs
    for i in range(col_num): # FIRST CHOICE
        for j in range(col_num): # SECOND CHOICE
            if(i == j): # SAMPLE VAR. CASE
                # get the average
                xbar = (1/row_num)*np.sum(X[:,i])
                cov_matrix[i,i] = (1/(row_num-1))*sum((X[:,i]-xbar)**2)  # Assign the SAMPLE var. to the main diagonal position, MAKE YOUR OWN VARIANCE FUNCTION
                # note: I am dumb and accidentally made the population variance algorithm at some point instead of the SAMPLE variance.
            else:
                # get the averages of xi and xj
                xbar_i = (1/row_num)*np.sum(X[:,i])
                xbar_j = (1/row_num)*np.sum(X[:,j])
                
                # Construct algorithm for covariance between the two column arrays with the given formula above.
                cov_matrix[i,j] = (1/(row_num - 1))*sum((X[:,i] - xbar_i)*(X[:,j] - xbar_j))
                
    return cov_matrix # return the covariance matrix

    # YOUR CODE HERE
    # raise NotImplementedError()

In [8]:
## check your work
X = np.random.random((5,3))  # or make up some other matrix for X 
np.array_equal(np.round(np.cov(X.T),4), np.round(covariance_matrix(X),4)) 

# again, we're rounding because your "by-hand" computation will differ slightly from NumPy's internal computations


True

## Problem 4

The code below generates a NumPy array X of shape (12,8) filled with integers. Please don't change the seed.

In [9]:
import numpy as np

np.random.seed(10)
X = np.random.randint(10,size = (12,8)) 
X

array([[9, 4, 0, 1, 9, 0, 1, 8],
       [9, 0, 8, 6, 4, 3, 0, 4],
       [6, 8, 1, 8, 4, 1, 3, 6],
       [5, 3, 9, 6, 9, 1, 9, 4],
       [2, 6, 7, 8, 8, 9, 2, 0],
       [6, 7, 8, 1, 7, 1, 4, 0],
       [8, 5, 4, 7, 8, 8, 2, 6],
       [2, 8, 8, 6, 6, 5, 6, 0],
       [0, 6, 9, 1, 8, 9, 1, 2],
       [8, 9, 9, 5, 0, 2, 7, 3],
       [0, 4, 2, 0, 3, 3, 1, 2],
       [5, 9, 0, 1, 0, 1, 9, 0]])

(a) (2 points) Write a function called "swap" that takes as its input the array $X$ and two integers n and m ($ 1 \le n,m \le 12, n \neq m$) and returns the matrix X in which rows n and m have been swapped. 

In [10]:
# COMPLETE
# input is a given array 'X' and two integers n,m.
# swap the rows 'n' and 'm'
def swap(X,n,m):
    """swaps rows n and m of array X"""
    # assign to variables to make swapping easier? Make sure to account for row 0
    row_n = X[n-1,:].copy() # choice 'n' row
    row_m = X[m-1,:].copy() # choice 'm' row
    
    # Start replacing each row
    X[n-1,:] = row_m # swap out row 'n' with row 'm'
    X[m-1,:] = row_n # swap out row 'm' with row 'n'
    
    return X # return the new array 'X'

        
    # YOUR CODE HERE
    # raise NotImplementedError()

In [11]:
# check your work
np.array_equal(swap(X,1,2), np.array([[9, 0, 8, 6, 4, 3, 0, 4],
       [9, 4, 0, 1, 9, 0, 1, 8],
       [6, 8, 1, 8, 4, 1, 3, 6],
       [5, 3, 9, 6, 9, 1, 9, 4],
       [2, 6, 7, 8, 8, 9, 2, 0],
       [6, 7, 8, 1, 7, 1, 4, 0],
       [8, 5, 4, 7, 8, 8, 2, 6],
       [2, 8, 8, 6, 6, 5, 6, 0],
       [0, 6, 9, 1, 8, 9, 1, 2],
       [8, 9, 9, 5, 0, 2, 7, 3],
       [0, 4, 2, 0, 3, 3, 1, 2],
       [5, 9, 0, 1, 0, 1, 9, 0]]))  


True

(b) (2 points) Write a function called "sort_by_column" that takes as its input the array X and an integer k ($1\le k \le 8$) and sorts the rows of X by the values of column k. Numbers that were in the same row before, should still be in the same row. Don't worry about ties - if there are ties, I don't care how you order them. 

In [12]:
# COMPLETE
def sort_by_column(X,k):
    """ sort rows of X by values in column k"""
    # When col 'k' (1,8) gets sorted, the order the values change in the column change for all the rows.
    Y = X.copy() # make a copy of the original
    
    # Retrieve an array of just the ordered POSITIONS
    col_order = np.argsort(Y[:,k-1])
    for i in range(len(X[:,0])): # traverse by ordered positions and start swapping accordingly by what positions were changed in the column
        X[i,:] = Y[col_order[i],:]
    
    return X # return the newly sorted array 'X'

print(X)
print(sort_by_column(X,8))
    # YOUR CODE HERE
    # raise NotImplementedError()

[[9 0 8 6 4 3 0 4]
 [9 4 0 1 9 0 1 8]
 [6 8 1 8 4 1 3 6]
 [5 3 9 6 9 1 9 4]
 [2 6 7 8 8 9 2 0]
 [6 7 8 1 7 1 4 0]
 [8 5 4 7 8 8 2 6]
 [2 8 8 6 6 5 6 0]
 [0 6 9 1 8 9 1 2]
 [8 9 9 5 0 2 7 3]
 [0 4 2 0 3 3 1 2]
 [5 9 0 1 0 1 9 0]]
[[2 6 7 8 8 9 2 0]
 [6 7 8 1 7 1 4 0]
 [2 8 8 6 6 5 6 0]
 [5 9 0 1 0 1 9 0]
 [0 6 9 1 8 9 1 2]
 [0 4 2 0 3 3 1 2]
 [8 9 9 5 0 2 7 3]
 [9 0 8 6 4 3 0 4]
 [5 3 9 6 9 1 9 4]
 [6 8 1 8 4 1 3 6]
 [8 5 4 7 8 8 2 6]
 [9 4 0 1 9 0 1 8]]
